linphonePackages: update & refactor

Naxdy opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This PR refactors the way Linphone is built in nixpkgs and how its dependencies are organized:

- Since Linphone has a _lot_ of custom (vendored) libraries, this PR introduces a new scope `linphonePackages` that contains all of them.
- To streamline building dependencies derived from the `linphone-sdk` monorepo, we have a helper function `mkLinphoneDerivation`.
- Due to the extent of the refactor, it is done in two commits: First remove everything from nixpkgs, as builds are failing right now anyway, then (re-)init the new packages.
- Previous maintainers are listed as co authors in the init commit, and are retained in the new derivations. In addition, I have added myself as maintainer to all packages.
- All packages have been updated to the current (as of 2025-09-20) stable versions.
- Build processes and dependencies have been adjusted wherever it was necessary.
- Obsolete patches have been modified / removed, and new patches have been added where it was needed.
- Metadata (licenses, descriptions, homepages, etc.) has been updated / corrected.
- CMake config tests have been added to certain libraries where the build process had to be adjusted to be able to find them.

Closes #432401 
Closes #292373
Closes #219551 

---


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Naxdy committed:
linphonePackages: remove

Linphone is a SIP softphone application, which depends on many bespoke
and custom-patched libraries, collectively referred to as
`linphonePackages` in this commit.

The build process has been failing for a while on both stable and
unstable, but due to the way the packages are organized in nixpkgs,
updating them is quite cumbersome.

Therefore, in an effort to ensure future maintainability, we first
remove all existing packages from nixpkgs, to then be able to later add
them again within their own scope.

Naxdy committed:
linphonePackages: init

Co-authored-by: Jaakko Luttinen <jaakko.luttinen@iki.fi>
Co-authored-by: Thibault Lemaire <thibault.lemaire@protonmail.com>
Co-authored-by: 7c6f434c <7c6f434c@mail.ru>

Naxdy commented:
cc @philiptaron

Naxdy committed:
libosmoabis: use `linphonePackages`

Naxdy committed:
trx: use `linphonePackages`

Naxdy committed:
mslibc: use `linphonePackages`

Naxdy committed:
soci: migrate to `pkgs/by-name`

philiptaron reviewed (commented):
Thank you for this large PR! I can tell it's a lot of work.

I don't have any blocking comments, but I would like answers for the questions I have in this PR, and would appreciate it if you accepted some of the suggestions.

My only really substantial piece of feedback is that I don't know if `linphonePackages` really needs to exist. Pulling all the various vendored libraries into a scope is a really good idea. It can happen inside of `pkgs/by-name/li/linphone`, keeping the rules of `by-name`. After all, it'd be pretty unexpected to have other applications make use of linphone versions outside of the derivation, right? Other than plugins.

Naxdy commented on pkgs/applications/networking/instant-messengers/linphone/belcard/default.nix:
```diff
@@ -0,0 +1,36 @@
+}:
+mkLinphoneDerivation {
+  pname = "belcard";
+
```
This would necessitate ignoring the passed `version` in favor of `linphoneVersion` within `mkLinphoneDerivation`, which I think forms kind of an anti-pattern. Otherwise maintainers run the risk of accidentally specifying a version in a new package manually, which would result in a failed build.

Other similar scopes like KDE also don't have this, for examples search the repo for invocations of `mkKdeDerivation`.

ThibaultLemaire reviewed (approved):
Thank you very much for the hard work.

Being the only user of NixOS in the company, I have previously attempted to update the existing derivations for my own personal use, but my lack of skill in both CMake & Nix soon bogged me down to a halt. (Even though I could talk in-person with the people who wrote the CMake files.)

So I can gauge the effort you put in this PR, and on a personal level it means a lot. Thanks again.

Naxdy committed:
linphonePackages: remove thibaultlemaire as maintainer

As per wishes, see:
https://github.com/NixOS/nixpkgs/pull/444603#pullrequestreview-3261894742

7c6f434c commented:
I have lost my SIP usecases and so I am mostly acting as merger for this package, but in general all the code makes sense to me, thanks for the refactorings and updates!

philiptaron reviewed (approved):
Let's get the merge conflicts fixed up, then merge! 🚀

philiptaron commented on pkgs/applications/networking/instant-messengers/linphone/default.nix:
```diff
@@ -1,144 +1,24 @@
+    let
+      packages = lib.filterAttrs (name: value: value == "directory") (builtins.readDir ./.);
+    in
+    libsForQt5 // (lib.mapAttrs (name: value: self.callPackage ./${name} { }) packages);
```
That makes enormous sense.

philiptaron reviewed (approved):
Will build

philiptaron commented:
Both `osmo-hnbgw` and `msilbc` are broken on master due to merging CMake 4. In order to build this derivation, they need some attention 😭

Naxdy committed:
ilbc: fix build with cmake 4

Naxdy commented:
I was able to fix `ilbc` and pushed a commit for it, but `osmo-hnbgw` seems to be unrelated to the CMake update. However, the library that was impacted (`libosmoabis`) by the linphone refactor does build, and given that `osmo-hnbgw` also doesn't build on master, I'd say it's unrelated. Here's an excerpt from the build logs:

```
hnbgw_cn.c: In function 'resolve_addr_name':
hnbgw_cn.c:556:56: error: invalid use of undefined type 'struct osmo_ss7_instance'
  556 |                      " cs7 instance %d != %d\n", (*ss7)->cfg.id, ss7_tmp->cfg.id);
      |                                                        ^~
hnbgw_cn.c:556:73: error: invalid use of undefined type 'struct osmo_ss7_instance'
  556 |                      " cs7 instance %d != %d\n", (*ss7)->cfg.id, ss7_tmp->cfg.id);
      |                                                                         ^~
hnbgw_cn.c:556:56: error: invalid use of undefined type 'struct osmo_ss7_instance'
  556 |                      " cs7 instance %d != %d\n", (*ss7)->cfg.id, ss7_tmp->cfg.id);
      |                                                        ^~
hnbgw_cn.c:556:73: error: invalid use of undefined type 'struct osmo_ss7_instance'
  556 |                      " cs7 instance %d != %d\n", (*ss7)->cfg.id, ss7_tmp->cfg.id);
      |                                                                         ^~
hnbgw_cn.c:556:56: error: invalid use of undefined type 'struct osmo_ss7_instance'
  556 |                      " cs7 instance %d != %d\n", (*ss7)->cfg.id, ss7_tmp->cfg.id);
      |                                                        ^~
hnbgw_cn.c:556:73: error: invalid use of undefined type 'struct osmo_ss7_instance'
  556 |                      " cs7 instance %d != %d\n", (*ss7)->cfg.id, ss7_tmp->cfg.id);
      |                                                                         ^~
hnbgw_rua.c: In function 'rua_to_scu':
hnbgw_rua.c:283:25: warning: 'ranap_cn_rx_co_decode' is deprecated: Use ranap_cn_rx_co_decode2() instead, the first arg is not used [-Wdeprecated-declarations]
  283 |                         rc = ranap_cn_rx_co_decode(map, message, msgb_l2(prim->oph.msg), msgb_l2len(prim->oph.msg));
      |                         ^~
hnbgw_cn.c:556:56: error: invalid use of undefined type 'struct osmo_ss7_instance'
  556 |                      " cs7 instance %d != %d\n", (*ss7)->cfg.id, ss7_tmp->cfg.id);
      |                                                        ^~
In file included from hnbgw_rua.c:46:
/nix/store/ca6fdbgcnb1vkjvs7d91fx6nr3izm1la-osmo-iuh-1.6.0/include/osmocom/ranap/ranap_common_cn.h:24:5: note: declared here
   24 | int ranap_cn_rx_co_decode(void *unused, ranap_message *message, uint8_t *data, size_t len)
      |     ^~~~~~~~~~~~~~~~~~~~~
hnbgw_cn.c:556:73: error: invalid use of undefined type 'struct osmo_ss7_instance'
  556 |                      " cs7 instance %d != %d\n", (*ss7)->cfg.id, ss7_tmp->cfg.id);
      |                                                                         ^~
hnbgw_cn.c: In function 'hnbgw_cnlink_init':
hnbgw_cn.c:598:20: error: invalid use of undefined type 'struct osmo_ss7_instance'
  598 |                 ss7->cfg.primary_pc = (23 << 3) + 5;
      |                    ^~
hnbgw_cn.c:601:38: error: invalid use of undefined type 'struct osmo_ss7_instance'
  601 |         if (!osmo_ss7_pc_is_valid(ss7->cfg.primary_pc)) {
      |                                      ^~
hnbgw_cn.c:603:25: error: invalid use of undefined type 'struct osmo_ss7_instance'
  603 |                      ss7->cfg.id);
      |                         ^~
hnbgw_cn.c:603:25: error: invalid use of undefined type 'struct osmo_ss7_instance'
  603 |                      ss7->cfg.id);
      |                         ^~
hnbgw_cn.c:603:25: error: invalid use of undefined type 'struct osmo_ss7_instance'
  603 |                      ss7->cfg.id);
      |                         ^~
hnbgw_cn.c:603:25: error: invalid use of undefined type 'struct osmo_ss7_instance'
  603 |                      ss7->cfg.id);
      |                         ^~
hnbgw_cn.c:606:23: error: invalid use of undefined type 'struct osmo_ss7_instance'
  606 |         local_pc = ss7->cfg.primary_pc;
      |                       ^~
hnbgw_cn.c:611:68: error: invalid use of undefined type 'struct osmo_ss7_instance'
  611 |         gw->sccp.client = osmo_sccp_simple_client_on_ss7_id(gw, ss7->cfg.id, "OsmoHNBGW",
      |                                                                    ^~
```

Unfortunately I'm not familiar with the project, so I'm not sure what exactly the issue is. A quick glance at the [release tags](https://github.com/osmocom/osmo-hnbgw/tags) suggest that it's been out of date for a while. Given the lack of maintainers, we could think about removing it, but that'd probably best happen in a separate PR.

philiptaron commented:
Thanks again for this high-quality work. Looking forward to seeing what comes out of the next release!

presto8 commented:
Awesome work, thank you! This allows me to use the recording feature of linphone, which I have not been able to up until now.

felixsinger commented:
Awesome. Thank you!
